3scale/3scale-operator
# THREESCALE-10525 - Tenant CR - extend capabilities

Jira: https://issues.redhat.com/browse/THREESCALE-10525 - Tenant CR - extend capabilities

_This task will be tested with new Porta-client task/PR:
        -     Jira: https://issues.redhat.com/browse/THREESCALE-10672
        -    [porta-client PR 59](https://github.com/3scale/3scale-porta-go-client/pull/59)_
01.Feb.24: **go.mod is updated to point to new Porta client tag: v0.10.0**

## What was done in PR:
- Added Info fields to Tenant CR.

## Validation Notes
We are testing that following INFO optional parameters can be set and updated in 3scale Tenant via Tenant CR:
      - **fromEmail**
      - **supportEmail**
      - **financeSupportEmail**
      - **siteAccessCode**
 _Note:   stateEvent - decided not to add, so removed from list_

- Get current branch, Install and run 3scale locally:
```sh
make install
oc new-project 3scale-test
make download
make run
```
- On another window:
All files examples could be found below.
```sh
oc apply -f s3-creds-secret.yaml
oc apply -f apimanagerCR.yaml
```
- Apply following secret, that required for tenant CR:
```
kind: Secret
apiVersion: v1
metadata:
  name: project-test-secret
  namespace: 3scale-test
data:
  admin_password: YWRtaW4K
type: Opaque
```
- Apply tenant CR yaml file that does NOT contains Info fields

```
apiVersion: capabilities.3scale.net/v1alpha1
kind: Tenant
metadata:
  generation: 3
  name: example1
  namespace: 3scale-test
spec:
  email: vmo@kuku.com
  masterCredentialsRef:
    name: system-seed
    namespace: 3scale-test
  organizationName: myOrg2
  passwordCredentialsRef:
    name: project-test-secret
    namespace: 3scale-test
  systemMasterUrl: https://master.xxx.xxxx.xxxx.xx.xx.org
  tenantSecretRef:
    name: system-seed
    namespace: 3scale-test
  username: tenant2FromCR1

```

```
oc apply -f tenantCR1.yaml
```

- Check that Tenant created in 3scale
    - Open 3scale master portal
        - Open 3scale API Docs -> Tenants section
        - Run Tenant GET (Show a tenant account) API 
![Screenshot from 2024-02-04 10-08-42](https://github.com/3scale/3scale-operator/assets/56625811/127660cd-4007-4e5e-af71-708da311a9f0)


  Output like following is expected
  We can see here Info fields, but they have  default values that were set by PORTA

```
<signup>
  <account>
    <id>6</id>
    <created_at>2024-02-04T08:06:03Z</created_at>
    <updated_at>2024-02-04T08:06:24Z</updated_at>
    <state>approved</state>
    <admin_domain>myorg2-2-admin.xxx.xx.xx.xx.xxx.org</admin_domain>
    <domain>myorg2-2.xxxx.xxx.xxx.xx.xxx.org</domain>
    <admin_base_url>https://myorg2-2-admin.xxx.xxxx.xx.xxx.xx.org</admin_base_url>
    <base_url>https://myorg2-2.xxx.xxx.xxx.xx.xxx.org</base_url>
    <from_email>no-reply@apps.xxx.xxx.xx.xxx.org</from_email>
    <support_email>vmo@kuku.com</support_email>
    <finance_support_email>vmo@kuku.com</finance_support_email>
    <site_access_code>xxxxxxx</site_access_code>
    <org_name>myOrg2</org_name>
    <extra_fields/>
    <monthly_billing_enabled>true</monthly_billing_enabled>
    <monthly_charging_enabled>true</monthly_charging_enabled>
    <credit_card_stored>false</credit_card_stored>
    <plans>
    .......
    .......
```

- Update tenant Info parameters in 3scale Tenant via CR
    Apply/Update CR - set Info parameters

```
apiVersion: capabilities.3scale.net/v1alpha1
kind: Tenant
metadata:
  generation: 3
  name: example1
  namespace: 3scale-test
spec:
  email: vmo@kuku.com
  masterCredentialsRef:
    name: system-seed
    namespace: 3scale-test
  organizationName: myOrg2
  passwordCredentialsRef:
    name: project-test-secret
    namespace: 3scale-test
  systemMasterUrl: https://master.xxxx.xx.xx.xx.xxx.org
  tenantSecretRef:
    name: system-seed
    namespace: 3scale-test
  username: tenant2FromCR1
  fromEmail: tenant2From@kuku.com
  supportEmail: tenant2Support@kuku.com
  financeSupportEmail: tenant2FinanceSupport@kuku.com
  siteAccessCode: abc123

```

- Check that 3scale Tenant was updated
   - Open 3scale master portal
        - Open 3scale API Docs -> Tenants section
        - Run Tenant GET (Show a tenant account) API

  Output like following is expected, that contains updated Info Parameters

```
<signup>
  <account>
    <id>6</id>
    <created_at>2024-02-04T08:06:03Z</created_at>
    <updated_at>2024-02-04T08:22:02Z</updated_at>
    <state>approved</state>
    <admin_domain>myorg2-2-admin.xxxx.xxxx.xxx.xxx.xx.org</admin_domain>
    <domain>myorg2-2.xxxx.xxxx.xxx.xxx.xx.org</domain>
    <admin_base_url>https://myorg2-2-admin.xxxx.xxxx.xxx.xxx.xx.org</admin_base_url>
    <base_url>https://myorg2-2.xxxx.xxxx.xxx.xxx.xx.org</base_url>
    <from_email>tenant2From@kuku.com</from_email>
    <support_email>tenant2Support@kuku.com</support_email>
    <finance_support_email>tenant2FinanceSupport@kuku.com</finance_support_email>
    <site_access_code>abc123</site_access_code>
    <org_name>myOrg2</org_name>
    <extra_fields/>
    <monthly_billing_enabled>true</monthly_billing_enabled>
    <monthly_charging_enabled>true</monthly_charging_enabled>
    <credit_card_stored>false</credit_card_stored>
.....
```

We can see here that site access code and emails changed
